src/librustc_mir/interpret/intern.rs

```diff
@@ -341,12 +340,19 @@ pub fn intern_const_alloc_recursive<M: CompileTimeMachine<'mir, 'tcx>>(
             // We can't call the `intern_shallow` method here, as its logic is tailored to safe
             // references and a `leftover_allocations` set (where we only have a todo-list here).
             // So we hand-roll the interning logic here again.
-            if base_intern_mode != InternMode::Static {
-                // If it's not a static, it *must* be immutable.
-                // We cannot have mutable memory inside a constant.
-                // FIXME: ideally we would assert that they already are immutable, to double-
-                // check our static checks.
-                alloc.mutability = Mutability::Immutable;
+            match base_intern_mode {
+                InternMode::Static => {},
+                InternMode::Const | InternMode::ConstBase => {
+                    // If it's not a static, it *must* be immutable.
+                    // We cannot have mutable memory inside a constant.
+                    // We use `delay_span_bug` here, because this can be reached in the presence
+                    // of fancy transmutes.
+                    if alloc.mutability == Mutability::Mutable {
+                        // For better errors later, mark the allocation as immutable
```

Comment:
Why just for "better errors"? Isn't this a crucial part of our safety checks, affects codegen, and also telling Miri that these must not be mutated any more?

Reply:
Since we are `delay_span_bug`ing out two lines below, it's entirely irrelevant what value this has. But we'll get more errors in later checks if we don't do this.

Reply:
Yeah I didn't see the `delay_span_bug` when writing this.